### DISCUSSION TITLE
hi, tigerj , the host should  install hplip driver for usb hp printer first?

### DISCUSSION THREAD

**Discussion by hawkinchina** [RagingTiger/docker-cups-airprint#5] [Q&A]
hi,tigerj:
    Thanks for your docker container. I will use this docker for arm64.  I have one question:
the host(linux) should  install hplip driver for usb hp printer first?  According to hp printer web, it suggest use hplip to drive this hp printer.

**Comment by RagingTiger** [comment]
tl;dr: **NO**

**SHORT EXPLANATION**
If you need the `printer drivers` found in the [`hplip`](https://en.wikipedia.org/wiki/HP_Linux_Imaging_and_Printing) package, then the [`tigerj/cups-airprint`](https://hub.docker.com/r/tigerj/cups-airprint) `Docker image` is built with the `hplip` package installed (see [Dockerfile](https://github.com/RagingTiger/docker-cups-airprint/blob/00c2129cec361f3c63822be9c93773e5a3d1204e/Dockerfile#L7)).

**LONG EXPLANATION**
`Docker` is great because you can simply **load** everything into a `Docker image` at build time, and then deploy it anywhere that `Docker` runs (assuming the architecture of the build and deploy environments are the same). This includes `drivers`.

In the [`Dockerfile`](https://github.com/RagingTiger/docker-cups-airprint/blob/00c2129cec361f3c63822be9c93773e5a3d1204e/Dockerfile#L7) for this repository, we are installing the [`hplip`](https://en.wikipedia.org/wiki/HP_Linux_Imaging_and_Printing) package (a commonly used package with thousands of `printer drivers` for `HP Printers`). Hence, we **do not** need to worry about our `host OS` having the required drivers installed .... so long as our application is running in a `Docker container`, generated from a `Docker image` that has the drivers installed (as is the case for the [`Dockerfile`](https://github.com/RagingTiger/docker-cups-airprint/blob/00c2129cec361f3c63822be9c93773e5a3d1204e/Dockerfile#L7) in this repo).

Now, that being said, the only time you would worry about the `host OS` having the required `drivers` installed is if you were going to **mount** a `host directory` into a `Docker container` using a [`Docker volume`](https://docs.docker.com/storage/volumes/). This would in essence allow you to use the `host OS` drivers **inside** the `Docker container` that your application is running in .... which would eliminate the need to install the drivers into the `Docker image` at build time.
